keps/sig-network/3766-referencegrant/README.md

```diff
@@ -0,0 +1,492 @@
+
+With all of that in mind, we felt that the benefits in terms of specificity
+outweighed the costs in terms of user pain to use the more correct `Resource`.
+
+### API Spec
+
+```golang
+// ReferenceGrant identifies kinds of resources in other namespaces that are
+// trusted to reference the specified kinds of resources in the same namespace
+// as the policy.
+//
+// Each ReferenceGrant can be used to represent a unique trust relationship.
+// Additional ReferenceGrants can be used to add to the set of trusted
+// sources of inbound references for the namespace they are defined within.
+//
+// ReferenceGrant is a form of runtime verification allowing users to assert
+// which cross-namespace object references are permitted. Implementations that
+// support ReferenceGrant MUST NOT permit cross-namespace references which have
+// no grant, and MUST respond to the removal of a grant by revoking the access
+// that the grant allowed.
+//
+// Support: Core
+type ReferenceGrant struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of ReferenceGrant.
+	Spec ReferenceGrantSpec `json:"spec,omitempty"`
+
+	// Note that `Status` sub-resource has been excluded at the
+	// moment as it was difficult to work out the design.
+	// A `Status` sub-resource may be added in the future.
+}
+
+// +kubebuilder:object:root=true
+// ReferenceGrantList contains a list of ReferenceGrant.
+type ReferenceGrantList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []ReferenceGrant `json:"items"`
+}
+
+// ReferenceGrantSpec identifies a cross namespace relationship that is trusted
+// for Gateway API.
+type ReferenceGrantSpec struct {
+	// From describes the trusted namespaces and kinds that can reference the
+	// resources described in "To". Each entry in this list MUST be considered
+	// to be an additional place that references can be valid from, or to put
+	// this another way, entries MUST be combined using OR.
+	//
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=16
+	From []ReferenceGrantFrom `json:"from"`
+
+	// To describes the resources that may be referenced by the resources
+	// described in "From". Each entry in this list MUST be considered to be an
+	// additional place that references can be valid to, or to put this another
+	// way, entries MUST be combined using OR.
+	//
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=16
+	To []ReferenceGrantTo `json:"to"`
+}
+
+// ReferenceGrantFrom describes trusted namespaces and kinds.
+type ReferenceGrantFrom struct {
+	// Group is the group of the referent.
+	// When empty, the Kubernetes core API group is inferred.
+	Group Group `json:"group"`
+
+	// Kind is the kind of the referent.
+	Kind string `json:"kind"`
+
+	// Namespace is the namespace of the referent.
+	Namespace string `json:"namespace"`
```

Comment:
Are we sure we do not want a label selector here?  `kubernetes.io/metadata.name` can be used to target a single namespace.  Mutating namespaces in any way is considered a privileged action, so why is this limited to a hard coded name?

Reply:
I'm on the fence with namespace selectors... that makes evaluation of the applicability of a reference grant significantly more difficult (have to read/watch changes to namespace labels, etc). I'd be curious to hear the experience of using namespace selectors in network policy.

hoisting from [sig-auth discussion](https://docs.google.com/document/d/1woLGRoONE3EBVx-wTb4pvp4CI7tmLZ6lS26VTbosLKM/edit#bookmark=id.9q15hi5lip46):

is "allow from all namespaces" possible/desirable to model? I tend to agree namespace selectors make it possible to overgrant accidentally, but for places that intend to allow cluster-wide reference (e.g. reference secret or config map containing trust roots from any route), is something like `namespace: "*"` plausible

Reply:
network policy is built (and AdminNetPolicy continues to build) for a notion of "tenant" - defined as a set of NS labels.

Reply:
> is "allow from all namespaces" possible/desirable to model?

At some point there was a desire to allow TLS certs and other things that are "cluster global", so yes.

Reply:
Enabling namespace boundaries to be crossed already introduces some risk, but going a step further and making it easy to allow references from _all namespaces_ feels especially risky. I would rather require namespaces to be individually specified, but I may be too narrowly looking at this from the perspective of Gateway API use cases.

Reply:
I think that allowing access _from_ all namespaces is okay as long as you have to _explicitly opt in_.

It should require either a special value (Contour's TLSCertificateDelegation used `*` for `namespace` for this reason), or an explicit "allow from all namespaces" field, which could have BIG TEXT CALLOUTS about being careful.

Reply:
We also skipped a label selector here because of the watch complexity it adds for implementations - keeping this as a string per `From` entry means that only updates to the ReferenceGrant itself can cause changes in the auth outcome, whereas doing a selector means that both ReferenceGrant and Namespace label changes need to be watched.

I accept that this is not a good reason for keeping this around forever though. 😄

Reply:
The problem I have with this is that there's almost nothing else which works this way - most everything else is selectors.

You might go look at https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/2091-admin-network-policy, specifically `AdminNetworkPolicyPeer`.  I'm not saying we need to adopt all of this off the bat, but it does seem like some notion of "same tenant" is going to be common.

e.g. allow references to all Services from any namespace which has the same value of label "example.com/team" as I do.  Maybe the indirection is not needed ("same as me") but "any namespace where example.com/team = myteam" seems totally reasonable?

Reply:
>The problem I have with this is that there's almost nothing else which works this way - most everything else is selectors.

If we consider this API as being more similar to RBAC than NetworkPolicy, a `*` would at least be consistent with the existing authorization APIs Kubernetes supports.

If we adopted the additional `NamespaceResource` struct would we also need to have additional nesting in the YAML to enable the future expansion in a backwards compatible way? 

Although we could make the case that a selector would be useful in place of both the namespace and resource names, I'm concerned about how this would increase complexity for both implementations and users. Leaving a path to allow that in the future is probably smart, but it would be unfortunate to leave an extra level of nesting in the API if it never ended up being needed.

Reply:
> would we also need to have additional nesting in the YAML to enable the future expansion in a backwards compatible way?

Yes.  In general you want a one-of to be a distinct struct, which means extra indent.

> it would be unfortunate to leave an extra level of nesting in the API if it never ended up being needed

tradeoffs, tradeoffs.. :)

Reply:
Moved to "open questions" since I think we're going to need more discussion here.